astropy/utils/data.py

```diff
@@ -983,24 +983,42 @@ def connect_ftp(self, user, passwd, host, port, dirs, timeout):
 
 def _download_file_from_source(source_url, show_progress=True, timeout=None,
                                remote_url=None, cache=False, pkgname='astropy',
-                               http_headers=None, ftp_tls=False):
+                               http_headers=None, ftp_tls=None):
     from astropy.utils.console import ProgressBarOrSpinner
 
+    if timeout == 0:
+        raise urllib.error.URLError(
+            f"URL {remote_url} was supposed to be downloaded but timeout was set to 0; "
+            f"if this is unexpected check the astropy config file for the option "
+            f"remote_timeout")
+
     if remote_url is None:
         remote_url = source_url
     if http_headers is None:
         http_headers = {}
 
+    if ftp_tls is None and urllib.parse.urlparse(remote_url).scheme == "ftp":
+        try:
+            return _download_file_from_source(source_url,
+                                              show_progress=show_progress,
+                                              timeout=timeout,
+                                              remote_url=remote_url,
+                                              cache=cache,
+                                              pkgname=pkgname,
+                                              http_headers=http_headers,
+                                              ftp_tls=False)
+        except urllib.error.URLError as e:
+            if e.reason.startswith("ftp error: error_perm"):
+                ftp_tls = True
```

Comment:
I am still not sure about this as `except` is expensive but it is now the default path for IERS via FTP. Is this refactoring absolutely necessary in the scope of this PR?

Reply:
I do not like adding the `ftp_tls` option to `download_file` (and not any of the other functions that call `download_file`) in a long-term maintenance release, mostly because it is a difficult option to use correctly - you have to know whether the remote FTP server expects TLS or not and if you get it wrong you get a hard-to-understand failure. (`wget`, by the way, uses `ftps` for URLs that involve FTP over TLS and simply fails if an `ftp` url points to a server that requires TLS; sadly `urllib` does not understand `ftps` and it's not an IANA standard.) #9964 doesn't have a solution for knowing whether to use TLS, it simply assumes TLS is required for all IERS-containing URLs, which will break if someone wants to use a non-TLS mirror. So it seemed to me that forcing the user to know whether the server supported TLS was bad API design - even `astropy` can't get it right internally - and we should not put it in a release.

Unfortunately, the cost of `except` is negligible here because TLS-requiring servers need an extra network round trip - we need to try FTP/no TLS and be refused. This is a problem intrinsic to "implicit FTPS" - it requires users to guess whether the server wants TLS or not. There is an "explicit FTP over TLS" that negotiates TLS, but apparently that is not what the NASA servers are using. My reasoning was that TLS requires several extra network round trips anyway so the extra cost of using an unannounced FTP over TLS server wasn't that big a deal.

Reply:
This can be moved to another PR; it is in fact separate from the cache format, though that PR will depend on this one (and I don't remember how that is supposed to work).

Reply:
Yes, you raised a good point. The `except` overhead is probably negligible all things considered and there is no reason to download it that often anyway.

While technically out of scope, I also don't see the point of dragging it out as a separate PR unless we hear objection from @eteq . If it drags out and the `ftp_tls` interface is released, then it is a lot more work to go through the deprecation cycle.

So, let's leave it here for now. Thanks!

Reply:
I have no heard back from Erik, so I'll have to assume he has no objection. If if I assume wrong, we can follow-up as a separate PR.